relevanceai/operations_new/label/base.py

```diff
@@ -0,0 +1,158 @@
+            str="cosine",
+        similarity_threshold : float
+            float=0.1,
+        label_field, optional
+            the field in the label documents that contains the label
+        label_vector_field, optional
+            the field in the label documents that contains the vector
+
+        Returns
+        -------
+            A list of documents with the field "_label_" set to the list of labels
+
+        """
+        # for each document
+        # get vector
+        # match vector against label vectors
+        # store labels
+        # return labelled documents
+
+        # Get all vectors
+        vectors = self.get_field_across_documents(vector_field, documents)
+        for i, vector in enumerate(vectors):
+            # search across
+            labels = self._get_nearest_labels(
+                vector,
+                label_field=label_field,
+                label_documents=label_documents,
+                expanded=expanded,
+                label_vector_field=label_vector_field,
+            )
+            # TODO: add inplace=True
+            self.set_field("_label_", documents[i], labels)
+        return documents
+
+    def _get_nearest_labels(
+        self,
+        vector,
+        label_documents,
+        label_field: str = "label",
+        expanded: bool = True,
+        label_vector_field: str = "label_chunkvector_",
+        similarity_metric: str = "cosine",
+        max_number_of_labels: int = 1,
+        similarity_threshold: float = 0.1,
+    ):
+        # perform cosine similarity
+        if similarity_metric == "cosine":
+            labels = self.cosine_similarity(
+                query_vector=vector,
+                vector_field=label_vector_field,
+                documents=label_documents,
+                max_number_of_labels=max_number_of_labels,
+                similarity_threshold=similarity_threshold,
+            )
+        else:
+            raise ValueError(
+                "Only cosine similarity metric is supported at the moment."
+            )
+
+        # for the label vectors
+        if expanded:
+            return labels
+        else:
+            # get a list of labels
+            return self.get_field_across_documents(label_field, labels)
+
+    def cosine_similarity(
+        self,
+        query_vector,
+        vector_field,
+        documents,
+        reverse=True,
+        score_field: str = "_label_score",
+        max_number_of_labels: int = 1,
+        similarity_threshold: float = 0,
+    ):
+        from scipy.spatial import distance
+
+        sort_key = [
+            1 - distance.cosine(i, query_vector)
+            for i in self.get_field_across_documents(vector_field, documents)
+        ]
+        self.set_field_across_documents(score_field, sort_key, documents)
+        labels = sorted(documents, reverse=reverse, key=lambda x: x[score_field])[
+            :max_number_of_labels
+        ]
+        labels = deepcopy(labels)
```

Comment:
whats the reason for this line?

Reply:
Because in the following lines we pop the labels - this happens inplace without a deepcopy. open to alternative solutions though